### PR TITLE
driver: display: Add support for bbc:microbit v2

### PR DESCRIPTION
**Pull request by LingaoM**
Add support display driver for new bbc:microbit v2 board.

This PR decouples the dependence of logic and hardware,
and has achieved maximum code reuse.

Signed-off-by: Lingao Meng <menglingao@xiaomi.com>

### CODE REVIEW

**Comment by carlescufi** [comment]
@LingaoM this looks mostly fine to me, I have a similar version myself, but didn't we say we would convert the driver to be a generic led-matrix display that reads its pin mappings from Devicetree?

**Comment by LingaoM** [comment]
> @LingaoM this looks mostly fine to me, I have a similar version myself, but didn't we say we would convert the driver to be a generic led-matrix display that reads its pin mappings from Devicetree?
i
Sorry, i don't understand what you mean, or can you provide an example?

File: samples/bluetooth/mesh_demo/src/board.h [hunk 1]
```diff
@@ -14,7 +14,8 @@ void board_button_1_pressed(void);
 uint16_t board_set_target(void);
 void board_play(const char *str);
 
-#if defined(CONFIG_BOARD_BBC_MICROBIT)
+#if defined(CONFIG_BOARD_BBC_MICROBIT) || \
+	defined(CONFIG_BOARD_BBC_MICROBIT_V2)
```
Comment (joerchan) [inline samples/bluetooth/mesh_demo/src/board.h:18]: Couldn't `CONFIG_MICROBIT_DISPLAY` be used instead? For all instances of this OR expression

File: samples/bluetooth/mesh/CMakeLists.txt [hunk 2]
```diff
@@ -12,8 +12,6 @@ project(mesh)
 
 target_sources(app PRIVATE src/main.c)
 
-if (BOARD STREQUAL bbc_microbit)
+if(CONFIG_MICROBIT_DISPLAY)
   target_sources(app PRIVATE src/microbit.c)
-else()
-  target_sources(app PRIVATE src/board.c)
```
Comment (joerchan) [inline samples/bluetooth/mesh/CMakeLists.txt:18]: Is board.c a file that is not used anymore? Shouldn't it be removed in that case, or is this change i mistake?

**Comment by carlescufi** [comment]
@vanwinkeljan and @jfischer-no and @gmarull should we accept this driver or else require the author to write a generic LED matrix display driver?

**Comment by gmarull** [comment]
I think that drivers not implementing any Zephyr API interface should go into `drivers/misc` for now. The same rationale was applied for the FT800 driver.

**Comment by jfischer-no** [comment]
> I think that drivers not implementing any Zephyr API interface should go into `drivers/misc` for now. The same rationale was applied for the FT800 driver.

That can be done later, PR is open for long time let go this in if the code is acceptable.

**Comment by gmarull** [comment]
> That can be done later, PR is open for long time let go this in if the code is acceptable.
> edit: Unlike FT800, this driver can be modified to support display controller API.

So then I'd either move it to `misc` or make it conform to the display API.

**Comment by vanwinkeljan** [comment]
@carlescufi  As you stated yourself before I thought the idea was to make a generic LED matrix driver out of this driver.

@gmarull 
> So then I'd either move it to misc or make it conform to the display API.

Agree, and if a generic LED matrix driver is made out of this driver it is better to line it up with the display API driver

**Comment by jfischer-no** [comment]
> @carlescufi As you stated yourself before I thought the idea was to make a generic LED matrix driver out of this driver.
> 
> @gmarull
> 
> > So then I'd either move it to misc or make it conform to the display API.
> 
> Agree, and if a generic LED matrix driver is made out of this driver it is better to line it up with the display API

@LingaoM are you interested in working on a generic **_GPIO_** matrix driver? I do not think this PR will be accepted. You can also close it yourself.

**Comment by LingaoM** [comment]
> > @carlescufi As you stated yourself before I thought the idea was to make a generic LED matrix driver out of this driver.
> 
> > 
> 
> > @gmarull
> 
> > 
> 
> > > So then I'd either move it to misc or make it conform to the display API.
> 
> > 
> 
> > Agree, and if a generic LED matrix driver is made out of this driver it is better to line it up with the display API
> 
> 
> 
> @LingaoM are you interested in working on a generic **_GPIO_** matrix driver?

Of course, I still don’t know what you are talking about generic matrix driver. Is there any documentation?

**Comment by carlescufi** [comment]
> > > @carlescufi As you stated yourself before I thought the idea was to make a generic LED matrix driver out of this driver.
> > 
> > 
> > > 
> > 
> > 
> > > @gmarull
> > 
> > 
> > > 
> > 
> > 
> > > > So then I'd either move it to misc or make it conform to the display API.
> > 
> > 
> > > 
> > 
> > 
> > > Agree, and if a generic LED matrix driver is made out of this driver it is better to line it up with the display API
> > 
> > 
> > @LingaoM are you interested in working on a generic **_GPIO_** matrix driver?
> 
> Of course, I still don’t know what you are talking about generic matrix driver. Is there any documentation?

What @vanwinkeljan and @jfischer-no suggested in this thread here:
https://github.com/zephyrproject-rtos/zephyr/pull/31000#discussion_r551278051

So, instead of having a micro:bit specific driver, we would have a generic LED matrix display driver that takes the rows and cols from Devicetree, as well as some extra info, and is able to drive any LED matrix out there. Does this make sense?

**Comment by peterniebert** [comment]
> > > > @carlescufi As you stated yourself before I thought the idea was to make a generic LED matrix driver out of this driver.
> > > 
> > > 
> > > > 
> > > 
> > > 
> > > > @gmarull
> > > 
> > > 
> > > > 
> > > 
> > > 
> > > > > So then I'd either move it to misc or make it conform to the display API.
> > > 
> > > 
> > > > 
> > > 
> > > 
> > > > Agree, and if a generic LED matrix driver is made out of this driver it is better to line it up with the display API
> > > 
> > > 
> > > @LingaoM are you interested in working on a generic **_GPIO_** matrix driver?
> > 
> > 
> > Of course, I still don’t know what you are talking about generic matrix driver. Is there any documentation?
> 
> What @vanwinkeljan and @jfischer-no suggested in this thread here:
> [#31000 (comment)](https://github.com/zephyrproject-rtos/zephyr/pull/31000#discussion_r551278051)
> 
> So, instead of having a micro:bit specific driver, we would have a generic LED matrix display driver that takes the rows and cols from Devicetree, as well as some extra info, and is able to drive any LED matrix out there. Does this make sense?

IMHO it does not: the micro:bit v2 uses a 5x5 matrix where the geometrical and the wiring logic is the same, whereas micro:bit v1
maps the 5x5 leds to a 3x9 matrix. Generic code will be nice but in many cases inefficient. I agree to the original author that this 5x5 matrix of the microbit is a special case and IMHO should be treated as such.

I undertook a new implementation of the  driver for v2 - from scratch while preserving the API, also as an exercise for myself to see how to integrate low level code with zephyr - and the following observations are based on this:

matrix pinout:
The most annoying bit about porting the current driver to V2 is not so much that the matrix topology changed from 3x9 to 5x5, but that one of the pins is on port 1 while the other pins are on port 0. The original implementation just supposes that all pins are on the same port.  The next annoying aspect in porting the original driver is that you would have to use two gpio devices. This would just get messy. But unfortunately, Zephyr does not know about individual pins in the Nordic sense. A truly generic driver would have to treat each pin individually and would not be able to use port writes at all.

In my implementation, I went about this by treating one dimension of the matrix (which is fully on port 0) by port sets and clears, whereas I used the other dimension as individual pins. But individual pins require using the nrf gpio API really. So I decided to use the pins directly by the nrfx api

timing:

A driver for line refresh necessarily produces hundreds of interrupts per second for the display refresh and it should IMHO be implemented efficiently, read low level. The original implementation pollutes the kernel with hundreds of alarms per second, where as the Nrf has several hardware timers that go unused. I used the nrfx_timer API (configuration allows to choose which timer to avoid conflicts) to create periodic interrupts while the driver is active.

For reference, and for discussion, I attach my current implementation. You can see that I basically implement it with a C++ class and a singleton pattern. Of course, it would be easy to backport it to C if there is any added value to that (I do not see it, but that is just me). The aim was to have as few global symbols as possible. Obviously, all variables are static. I developed it within the original sample directory zephyr/samples/boards/bbc-microbit/display, I had to copy the font files into the src directory so I could develop without touching global configuration files. The result is here:

[display.zip](https://github.com/zephyrproject-rtos/zephyr/files/6636686/display.zip)

My guess is that it would be easier to backport this implementation to be v1 compatible than the other way around.

Thank you for your opinion.

-- Peter

**Comment by carlescufi** [comment]
@peterniebert thank you very much for the explanation and the code. I am not against using a specific nrf timer here, since this display is only ever going to work with a Nordic chip.
Could you please post an alternative PR with your code (ported from C++ to C) so we can take a look and see what makes sense for Zephyr upstream?

**Comment by peterniebert** [comment]
> @peterniebert thank you very much for the explanation and the code. I am not against using a specific nrf timer here, since this display is only ever going to work with a Nordic chip.
> Could you please post an alternative PR with your code (ported from C++ to C) so we can take a look and see what makes sense for Zephyr upstream?
OK, I'll do that, quick and dirty in order to get the discussion going. I think it will take several iterations before the things are right.

**Comment by peterniebert** [comment]
I submitted a new pull request for my implementation, see here :+1: 
https://github.com/zephyrproject-rtos/zephyr/pull/36264

**Comment by peterniebert** [comment]
> I submitted a new pull request for my implementation, see here +1
> #36264

Rather this one: https://github.com/zephyrproject-rtos/zephyr/pull/36280

**Pull request closed without merging**